src/main/java/com/gregtechceu/gtceu/api/data/chemical/material/ItemMaterialData.java

```diff
@@ -0,0 +1,198 @@
+    public static final Map<MaterialEntry, ArrayList<Supplier<? extends Block>>> MATERIAL_ENTRY_BLOCK_MAP = new Object2ObjectOpenHashMap<>();
+    /** Mapping of stone type blockState to "prefix, material" */
+    public static final Map<Supplier<BlockState>, TagPrefix> ORES_INVERSE = new Object2ObjectOpenHashMap<>();
+
+    public static final Map<ItemStack, List<ItemStack>> UNRESOLVED_ITEM_MATERIAL_INFO = new Object2ObjectOpenCustomHashMap<>(
+            ItemStackHashStrategy.comparingAllButCount());
+
+    public static void registerMaterialInfo(ItemLike item, ItemMaterialInfo materialInfo) {
+        if (item instanceof Block block) {
+            ITEM_MATERIAL_INFO.put(block, materialInfo);
+            return;
+        } else if (item instanceof BlockItem blockItem) {
+            ITEM_MATERIAL_INFO.put(blockItem.getBlock(), materialInfo);
+            return;
+        } else if (item instanceof ItemEntry<?> entry) {
+            ITEM_MATERIAL_INFO.put(entry.asItem(), materialInfo);
+            return;
+        }
+        ITEM_MATERIAL_INFO.put(item, materialInfo);
+    }
+
+    public static ItemMaterialInfo getMaterialInfo(ItemLike item) {
+        if (item instanceof Block block) {
+            return ITEM_MATERIAL_INFO.get(block);
+        } else if (item instanceof BlockItem blockItem) {
+            var info = ITEM_MATERIAL_INFO.get(blockItem.getBlock());
+            if (info != null) return info;
+            return ITEM_MATERIAL_INFO.get(item);
+        } else if (item instanceof ItemEntry<?> entry) {
+            return ITEM_MATERIAL_INFO.get(entry.asItem());
+        }
+        return ITEM_MATERIAL_INFO.get(item);
+    }
+
+    public static void clearMaterialInfo(ItemLike item) {
+        if (item instanceof Block block) {
+            ITEM_MATERIAL_INFO.remove(block);
+            return;
+        } else if (item instanceof BlockItem blockItem) {
+            var info = ITEM_MATERIAL_INFO.get(blockItem.getBlock());
+            if (info != null) {
+                ITEM_MATERIAL_INFO.remove(blockItem.getBlock());
+                return;
+            }
+            ITEM_MATERIAL_INFO.remove(item);
+            return;
+        } else if (item instanceof ItemEntry<?> entry) {
+            ITEM_MATERIAL_INFO.remove(entry.asItem());
+            return;
+        }
+        ITEM_MATERIAL_INFO.remove(item);
+    }
+
+    @SafeVarargs
+    public static void registerMaterialInfoItems(MaterialEntry materialEntry,
+                                                 Supplier<? extends ItemLike>... items) {
+        MATERIAL_ENTRY_ITEM_MAP.computeIfAbsent(materialEntry, entry -> new ArrayList<>())
+                .addAll(Arrays.asList(items));
+        for (Supplier<? extends ItemLike> item : items) {
+            ITEM_MATERIAL_ENTRY.add(Pair.of(item, materialEntry));
+            if (item instanceof Block block) {
```

Comment:
Since you're already going to loop over the items, you might as well not do the `Arrays.asList` call.
```suggestion
        var entryList = MATERIAL_ENTRY_ITEM_MAP.computeIfAbsent(materialEntry, entry -> new ArrayList<>());
        for (Supplier<? extends ItemLike> item : items) {
            ITEM_MATERIAL_ENTRY.add(Pair.of(item, materialEntry));
            entryList.add(item);
            if (item instanceof Block block) {
```